Modify tag list

lsq921224 opened the pull request:
Closes #13

rmacklin commented on app/views/images/_form.html.erb:
```diff
@@ -0,0 +1,7 @@
+<%= simple_form_for @image do |f| %>
```
Prefer using local variables rather than instance variables in partials; it makes them more easily reusable/less coupled to a particular view context.

rmacklin commented on app/views/images/_form.html.erb:
```diff
@@ -1,5 +1,5 @@
 <%= simple_form_for @image do |f| %>
-  <%= f.input :link %>
+  <%= f.input :link, disabled: (true if controller.action_name == 'edit') %>
```
Along the same lines as my [previous comment](https://github.com/AppFolioOnboarding/image-sharer-lsq921224/pull/49#discussion_r77931705), you should avoid having your partial inspect the environment in which its rendered. Instead, invert the dependency such that the callers who render this partial explicitly supply the information that the partial requires.

rmacklin commented on app/assets/stylesheets/_bootstrap-override.scss:
```diff
@@ -1,3 +1,7 @@
+.btn {
+  margin-right: 4px;
+}
```
This may be fine, but I just want to ask if this is something you truly want to apply to _all `.btn` elements throughout the application_ or if it was only added for some subset of these elements. If the latter, targeting this style more specifically would be preferred. As usual for CSS, I'll defer to @aaronpanch's opinion on the matter.

rmacklin commented on test/controllers/images_controller_test.rb:
```diff
@@ -199,4 +199,34 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
+    assert_equal %w(tag), image.reload.tag_list, 'Tag list is modified'
+  end
+
+  test "image's url should not be modified even if a user does an explicit http request" do
```
Great job testing this case!

rmacklin commented on test/controllers/images_controller_test.rb:
```diff
@@ -199,4 +199,34 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
+  test "image's url should not be modified even if a user does an explicit http request" do
+    image = Image.create!(image_valid_params)
+
+    put image_path(image), params: { image: { link: 'http://google.com/a.jpg', tag_list: 'new, tag' } }
```
The way this test is currently written is a bit brittle. If someone were to update `image_valid_params` to have the `link` set to `'http://google.com/a.jpg'`, this test would no longer ensure that the URL wasn't updated. If we are explicit about the URL the image is created with we can avoid this.

rmacklin replied:
Following up on this, I found the [`image_valid_params` method](https://github.com/AppFolioOnboarding/image-sharer-lsq921224/blob/7aa9c390ff8e5848650b2c4b709601684afcb02d/test/test_helper.rb#L39-L44) which I had not seen before. I see that method was added to DRY some of the `Image.create!` calls in your tests such that adding a new required attribute wouldn't require updating each of those. That's good but I think it would be better to extract the image creation itself into a private method that accepts arguments which would override the defaults:
```ruby
def create_image(link: 'http://google.com/a.jpeg', tag_list: 'tag, tag1, tag2')
  Image.create!(link: link, tag_list: tag_list)
end
```
This makes it easy to pass an explicit `link` and still have the valid default tags added for you.

Additionally, defining those methods directly in `ActiveSupport::TestCase` is an antipattern because it will make those methods available in all your tests, even where they aren't needed (e.g. none of your unit tests need to know how to log a user in). To avoid that, you can create modules to contain different test helper methods and `include` those modules in _only_ the tests that need them.

rmacklin commented on test/controllers/images_controller_test.rb:
```diff
@@ -199,4 +199,34 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
+    image = Image.create!(link: 'http://a.com/a.jpg', tag_list: 'tag')
+
+    put image_path(image), params: { image: { tag_list: '' } }
+    assert_response :unprocessable_entity
```
What else can be asserted about the response?

weixing2014 commented on app/controllers/images_controller.rb:
```diff
@@ -1,5 +1,5 @@
 class ImagesController < ApplicationController
-  before_action :find_image, only: [:destroy, :share, :send_mail]
+  before_action :find_image, only: [:destroy, :share, :send_mail, :update]
```
It seems functions `edit` and `show` need to call `find_image` as the first step of each function as well.

lsq921224 replied:
Well, there's a slight difference. `edit` and `show` calls `find`: 
`    @image = Image.find(params[:id])
`
but the others call `find_by`:
`    @image = Image.find_by(id: params[:id])
`
I use `find` for `edit` and `show` because rails will then take care of the `RecordNotFound` case for me.
the reason why I did not DRY `edit` and `show` is because if I do so, these two functions will have nothing in their body, and I think it reduces the readability. But since you mention it, I guess I will do it as well.

weixing2014 replied:
@lsq921224 Thanks for explaining between `find` and `find_by`! Also I did some research and it seemed there are some different opinions about [using `before_action`](http://craftingruby.com/posts/2015/05/31/dont-use-before-action-to-load-data.html). Then I think it's fine to leave it as is. Thanks!

weixing2014 commented on app/views/images/_form.html.erb:
```diff
@@ -0,0 +1,7 @@
+<%= simple_form_for @image do |f| %>
+  <%= f.input :link, disabled: (true if controller.action_name == 'edit') %>
```
If you really want to disable the link only if the action name is `edit`, maybe below is slightly easier to read:
```
<%= f.input :link, disabled: controller.action_name.eql? 'edit' %>
```

weixing2014 commented on test/page_objects/images/image_card.rb:
```diff
@@ -14,6 +14,11 @@ def go_to_share_page!
         window.change_to(SharePage)
       end
 
+      def go_to_edit_page!
```
I think we usually name this function `edit!` in APM.
@rmacklin should know more about it.

lsq921224 commented:
@rmacklin @weixing2014 I think I have addressed your comments except https://github.com/AppFolioOnboarding/image-sharer-lsq921224/pull/49#discussion_r77933018, for which I will open a separate PR to refactor my tests.

rmacklin commented on app/views/images/_form.html.erb:
```diff
@@ -1,5 +1,5 @@
 <%= simple_form_for image do |f| %>
-  <%= f.input :link %>
+  <%= f.input :link, disabled: controller.action_name.eql?('edit') %>
```
This comment still applies: https://github.com/AppFolioOnboarding/image-sharer-lsq921224/pull/49#discussion_r77931893

> you should avoid having your partial inspect the environment in which it's being rendered. Instead, invert the dependency such that the callers who render this partial explicitly supply the information that the partial requires.

rmacklin commented on app/views/images/_form.html.erb:
```diff
@@ -0,0 +1,7 @@
+<%= simple_form_for image do |f| %>
+  <%= f.input :link, disabled: controller.action_name.eql?('edit') %>
```
This comment still applies: https://github.com/AppFolioOnboarding/image-sharer-lsq921224/pull/49#discussion_r77931893

> you should avoid having your partial inspect the environment in which it's being rendered. Instead, invert the dependency such that the callers who render this partial explicitly supply the information that the partial requires.

lsq921224 replied:
Sorry that I missed it because Xing had another comment on that line. I have updated it.

rmacklin commented on app/views/images/_form.html.erb:
```diff
@@ -0,0 +1,7 @@
+<%= simple_form_for image do |f| %>
+  <%= f.input :link, disabled: disable_link %>
```
Minor but `disable_link_input` would be a little more clear (since `disable_link` could be misinterpreted as disabling a link on the page).

rmacklin replied:
That said, good job making this change in response to https://github.com/AppFolioOnboarding/image-sharer-lsq921224/pull/49#discussion_r78112197. However, I wanted to discuss an alternative. Do you think it's necessary to show this disabled input on the edit page, given that it's not editable, and we know which image we're editing because a preview is displayed on the edit page?

rmacklin replied:
If you think we can omit the `link` input on the edit page, I can use this opportunity to show you a cool feature of rendering partials that makes them more flexible: partial layouts.

Of course, one way to omit the `link` input would be with something like:
```erb
<% if show_link_input %>
  <%= f.input :link, disabled: disable_link %>
<% end %>
```

However, this makes the partial very inflexible. All you can do is either render the `link` input or not. In our case, we actually want to be able to render the `link` input or render a preview of the image. We can achieve this flexibility using partial layouts.

If you replace the `link` input with a `yield` in your partial, like:
```erb
<%= simple_form_for image do |f| %>
  <% yield f if block_given? %>

  <%= f.input :tag_list, input_html: { value: image.tag_list.join(',') } %>
  <div class="form-group row">
    <%= f.button :submit, class: 'btn-primary' %>
  </div>
<% end %>
```
You can render it as a partial layout. For the new page you'd pass a block that renders the `link` input:
```erb
<%= render layout: 'images/form', locals: { image: @image } do |form| %>
  <%= form.input :link %>
<% end %>
```
And for the edit page, you can pass a block that renders the image preview:
```erb
<%= render layout: 'images/form', locals: { image: @image } %>
  <div class="form-group row">
    <%= link_to image_path(@image), class: 'offset-sm-2' do %>
      <%= image_tag(@image.link, class: 'share-form-image') %>
    <% end %>
  </div>
<% end %>
```

This approach lets you reuse the pieces of the image form that are actually shared, while giving you the flexibility to insert anything else you want from the template that's rendering this partial.

rmacklin replied:
cc @weixing2014

weixing2014 replied:
Thanks, @rmacklin !

lsq921224 replied:
@rmacklin I really like this idea of  `<% yield f if block_given? %>`, but for some reason it's not showing the block I supply to it. i have changed my code to https://github.com/AppFolioOnboarding/image-sharer-lsq921224/pull/49#discussion_r78235833 exactly. I am trying to figure out why, but it would be great if you can give me some clue on that.

rmacklin replied:
It looks like a typo in my comment,
`<% yield f if block_given? %>`
should be
`<%= yield f if block_given? %>`
so that the block is actually rendered. Does that fix it?

lsq921224 replied:
yes, it works!

rmacklin commented on app/views/images/new.html.erb:
```diff
@@ -1,7 +1,3 @@
-    <%= f.button :submit, class: 'btn-primary' %>
-  </div>
+<%= render layout: 'images/form', locals: { image: @image } do |form| %>
+  <%= form.input :link %>
```
Nice!

rmacklin commented on app/views/images/edit.html.erb:
```diff
@@ -0,0 +1,7 @@
+  <%= link_to image_path(@image), class: 'offset-sm-2' do %>
+    <%= image_tag(@image.link, class: 'share-form-image') %>
+  <% end %>
+</div>
```
Minor: this should be indented a level

rmacklin replied:
Minor: this should be indented a level. Other than that, nice!

rmacklin commented:
> I think I have addressed your comments except https://github.com/AppFolioOnboarding/image-sharer-lsq921224/pull/49#discussion_r77933018, for which I will open a separate PR to refactor my tests.

I saw that PR (#51) and left comments there. Whichever is merged first will require changes in the other.

lsq921224 commented:
I fixed this https://github.com/AppFolioOnboarding/image-sharer-lsq921224/pull/49#discussion_r78458915 and I'm merging this PR, then I will make corresponding change in https://github.com/AppFolioOnboarding/image-sharer-lsq921224/pull/51
